From d1c5ed426e47b4cb661ef56dbc908a57ff92a823 Mon Sep 17 00:00:00 2001
From: Chris <meischris39@hotmail.com>
Date: Fri, 16 Oct 2026 08:39:15 +0100
Subject: [PATCH] [RetroPlayer] Only strip shader parameter pragmas that start
 a line

Kodi removes "#pragma parameter" lines before compiling a GLSL shader,
but it cut from those words wherever they appeared, up to and including
the newline. A comment that mentions them, or a commented-out parameter,
then took the next line with it. The TV Out shaders lost their parameter
uniforms this way and failed to compile.

Only remove lines that start with the pragma. A pragma on a file's last
line, with no newline after it, no longer hangs.
---
 cmake/treedata/common/tests.txt               |  1 +
 .../cores/RetroPlayer/shaders/ShaderUtils.cpp | 22 +++++---
 .../RetroPlayer/shaders/test/CMakeLists.txt   |  3 +
 .../shaders/test/TestShaderUtils.cpp          | 56 +++++++++++++++++++
 4 files changed, 75 insertions(+), 7 deletions(-)
 create mode 100644 xbmc/cores/RetroPlayer/shaders/test/CMakeLists.txt
 create mode 100644 xbmc/cores/RetroPlayer/shaders/test/TestShaderUtils.cpp

diff --git a/cmake/treedata/common/tests.txt b/cmake/treedata/common/tests.txt
index 3a03cc3..1d0b9d7 100644
--- a/cmake/treedata/common/tests.txt
+++ b/cmake/treedata/common/tests.txt
@@ -5,6 +5,7 @@ xbmc/cores/AudioEngine/Utils/test test/audioengine_utils
 xbmc/cores/RetroPlayer/playback/test test/retroplayer_playback
 xbmc/cores/RetroPlayer/rendering/test test/retroplayer_rendering
 xbmc/cores/RetroPlayer/savestates/test test/retroplayer_savestates
+xbmc/cores/RetroPlayer/shaders/test test/retroplayer_shaders
 xbmc/cores/RetroPlayer/streams/memory/test test/retroplayer_memory
 xbmc/cores/RetroPlayer/test test/retroplayer
 xbmc/cores/VideoPlayer/Buffers/test test/videoplayer_buffers
diff --git a/xbmc/cores/RetroPlayer/shaders/ShaderUtils.cpp b/xbmc/cores/RetroPlayer/shaders/ShaderUtils.cpp
index 8c414d3..b893cf1 100644
--- a/xbmc/cores/RetroPlayer/shaders/ShaderUtils.cpp
+++ b/xbmc/cores/RetroPlayer/shaders/ShaderUtils.cpp
@@ -8,19 +8,27 @@
 
 #include "ShaderUtils.h"
 
+#include <string_view>
+
 using namespace KODI::SHADER;
 
 std::string CShaderUtils::StripParameterPragmas(std::string source)
 {
-  size_t pragmaPosition;
-  size_t newlinePosition;
+  constexpr std::string_view PRAGMA = "#pragma parameter";
 
-  while ((pragmaPosition = source.find("#pragma parameter")) != std::string::npos)
+  // Only a line that starts with the pragma declares a parameter. The words
+  // can also appear in a comment, and the line after it must survive.
+  size_t lineStart = 0;
+  while (lineStart < source.size())
   {
-    newlinePosition = source.find_first_of('\n', pragmaPosition + 17);
-
-    if (newlinePosition != std::string::npos)
-      source.erase(pragmaPosition, newlinePosition - pragmaPosition + 1);
+    size_t lineEnd = source.find('\n', lineStart);
+    lineEnd = (lineEnd == std::string::npos) ? source.size() : lineEnd + 1;
+
+    const size_t textStart = source.find_first_not_of(" \t", lineStart);
+    if (textStart < lineEnd && source.compare(textStart, PRAGMA.size(), PRAGMA) == 0)
+      source.erase(lineStart, lineEnd - lineStart);
+    else
+      lineStart = lineEnd;
   }
 
   return source;
diff --git a/xbmc/cores/RetroPlayer/shaders/test/CMakeLists.txt b/xbmc/cores/RetroPlayer/shaders/test/CMakeLists.txt
new file mode 100644
index 0000000..d7a71ae
--- /dev/null
+++ b/xbmc/cores/RetroPlayer/shaders/test/CMakeLists.txt
@@ -0,0 +1,3 @@
+set(SOURCES TestShaderUtils.cpp)
+
+core_add_test_library(retroplayer_shaders_test)
diff --git a/xbmc/cores/RetroPlayer/shaders/test/TestShaderUtils.cpp b/xbmc/cores/RetroPlayer/shaders/test/TestShaderUtils.cpp
new file mode 100644
index 0000000..f3eb325
--- /dev/null
+++ b/xbmc/cores/RetroPlayer/shaders/test/TestShaderUtils.cpp
@@ -0,0 +1,56 @@
+/*
+ *  Copyright (C) 2026 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "cores/RetroPlayer/shaders/ShaderUtils.h"
+
+#include <string>
+
+#include <gtest/gtest.h>
+
+using namespace KODI::SHADER;
+
+TEST(TestShaderUtils, StripsAParameterLine)
+{
+  const std::string source = "#pragma parameter GAMMA \"Gamma\" 2.2 1.0 3.0 0.1\n"
+                             "uniform float GAMMA;\n";
+
+  EXPECT_EQ(CShaderUtils::StripParameterPragmas(source), "uniform float GAMMA;\n");
+}
+
+TEST(TestShaderUtils, StripsAnIndentedParameterLine)
+{
+  const std::string source = "  #pragma parameter GAMMA \"Gamma\" 2.2 1.0 3.0 0.1\n"
+                             "uniform float GAMMA;\n";
+
+  EXPECT_EQ(CShaderUtils::StripParameterPragmas(source), "uniform float GAMMA;\n");
+}
+
+TEST(TestShaderUtils, StripsAParameterOnTheLastLine)
+{
+  const std::string source = "uniform float GAMMA;\n"
+                             "#pragma parameter GAMMA \"Gamma\" 2.2 1.0 3.0 0.1";
+
+  EXPECT_EQ(CShaderUtils::StripParameterPragmas(source), "uniform float GAMMA;\n");
+}
+
+TEST(TestShaderUtils, KeepsAPragmaMentionedInAComment)
+{
+  const std::string source = "#ifdef PARAMETER_UNIFORM // Defined if #pragma parameters work.\n"
+                             "uniform float GAMMA;\n"
+                             "#endif\n";
+
+  EXPECT_EQ(CShaderUtils::StripParameterPragmas(source), source);
+}
+
+TEST(TestShaderUtils, KeepsACommentedOutParameter)
+{
+  const std::string source = "//#pragma parameter GAMMA \"Gamma\" 2.2 1.0 3.0 0.1\n"
+                             "uniform float GAMMA;\n";
+
+  EXPECT_EQ(CShaderUtils::StripParameterPragmas(source), source);
+}
-- 
2.53.0

